### PR TITLE
bug: no warning shown for parameterized conversions

### PR DESCRIPTION
**Pull request by vicente-romero-oracle**
no warning is shown for parameterized conversions

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.java.net/census#mcimadamore) (@mcimadamore - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/valhalla pull/666/head:pull/666` \
`$ git checkout pull/666`

Update a local copy of the PR: \
`$ git checkout pull/666` \
`$ git pull https://git.openjdk.java.net/valhalla pull/666/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 666`

View PR using the GUI difftool: \
`$ git pr show -t 666`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/valhalla/pull/666.diff">https://git.openjdk.java.net/valhalla/pull/666.diff</a>

</details>

### CODE REVIEW

**Comment by vicente-romero-oracle** [comment]
> I'm not too convinced by this fix. Using warnStack seems really a workaround, and, on top of my head, I can't think of other code doing this. IMHO, universal vars introduce a new relation between types: `isBoundedBy`. What you want here is, essentially, a form to do `isBoundedByUnchecked` which should do strict check, followed by a looser one (like isSubtypeUnchecked does). There are many ways to do this, but it feels that having a better separation between these different routines could be, in the long term, a simplifying move (even if, in the short term it will create some code duplication). Because then we can fix the callsites to use the relationship we think is correct, depending on the case.

I'm not sure that `isSubtype` / `isSubtypeUnchecked` is the right analogy here as in most cases we get to the place where the warning is being issue in this PR without going through `isBoundedBy`

**Comment by mcimadamore** [comment]
> > I'm not too convinced by this fix. Using warnStack seems really a workaround, and, on top of my head, I can't think of other code doing this. IMHO, universal vars introduce a new relation between types: `isBoundedBy`. What you want here is, essentially, a form to do `isBoundedByUnchecked` which should do strict check, followed by a looser one (like isSubtypeUnchecked does). There are many ways to do this, but it feels that having a better separation between these different routines could be, in the long term, a simplifying move (even if, in the short term it will create some code duplication). Because then we can fix the callsites to use the relationship we think is correct, depending on the case.
> 
> I'm not sure that `isSubtype` / `isSubtypeUnchecked` is the right analogy here as in most cases we get to the place where the warning is being issue in this PR without going through `isBoundedBy`

I got confused, sorry (too many type relations!). Looking at the code some more, it seems to me that what we want is to enhance `isSubtypeUnchecked`, right? E.g. `isSubtypeUnchecked` is already called in places where the compiler allows unchecked conversions. But we need to add a new kind of unchecked conversion, which requires some recursion (because isSubtypeUnchecked can go to containsType which then can go back to isSubtype - not unchecked!).

So, it seems to me that the subtyping routine has to be parameterized by the kind of containment routine we want to use. In the normal case, we'd just use Types::containsType - but if we want to check for nested unchecked warnings, we need to have a version of Types::containsType that is more flexible.

In other words, I think what we need is a Types::containsTypeUnchecked - which we can use as a parameter to the subtyping routine.

**Comment by vicente-romero-oracle** [comment]
> > > I'm not too convinced by this fix. Using warnStack seems really a workaround, and, on top of my head, I can't think of other code doing this. IMHO, universal vars introduce a new relation between types: `isBoundedBy`. What you want here is, essentially, a form to do `isBoundedByUnchecked` which should do strict check, followed by a looser one (like isSubtypeUnchecked does). There are many ways to do this, but it feels that having a better separation between these different routines could be, in the long term, a simplifying move (even if, in the short term it will create some code duplication). Because then we can fix the callsites to use the relationship we think is correct, depending on the case.
> > 
> > 
> > I'm not sure that `isSubtype` / `isSubtypeUnchecked` is the right analogy here as in most cases we get to the place where the warning is being issue in this PR without going through `isBoundedBy`
> 
> I got confused, sorry (too many type relations!). Looking at the code some more, it seems to me that what we want is to enhance `isSubtypeUnchecked`, right? E.g. `isSubtypeUnchecked` is already called in places where the compiler allows unchecked conversions. But we need to add a new kind of unchecked conversion, which requires some recursion (because isSubtypeUnchecked can go to containsType which then can go back to isSubtype - not unchecked!).
> 
> So, it seems to me that the subtyping routine has to be parameterized by the kind of containment routine we want to use. In the normal case, we'd just use Types::containsType - but if we want to check for nested unchecked warnings, we need to have a version of Types::containsType that is more flexible.
> 
> In other words, I think what we need is a Types::containsTypeUnchecked - which we can use as a parameter to the subtyping routine.

I will try this, I was going through other way, to untangle some of the type routines, the good thing is that I already have a separate containtType relation for the unchecked case so I will build from there

**Comment by vicente-romero-oracle** [comment]
first attempt at the type routines parameterization approach, what do you think?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java [hunk 1]
```diff
@@ -4864,7 +4946,7 @@ private boolean giveWarning(Type from, Type to) {
 
     private boolean containsTypeEquivalent(Type t, Type s) {
         return isSameType(t, s) || // shortcut
-            containsType(t, s) && containsType(s, t);
+            containsType(t, s, containsTypeUnchecked) && containsType(s, t, containsTypeUnchecked);
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:4949]: Is this the right replacement? It seems to me that the answer is "it depends". This is called both by "hasSameArgs" which is used to determine overriding (in which case a loose semantics could be ok), but also by the isSameType visitor - in which case the loose semantics should only be enabled conditionally.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java [hunk 2]
```diff
@@ -2029,6 +2029,12 @@ public boolean isReferenceProjection() {
         public boolean isValueProjection() {
             return isUniversal() && !isReferenceProjection();
         }
+
+        public Type withTypeVar(Type t) {
+            return t.hasTag(TYPEVAR) && t.isReferenceProjection() && t == projection ?
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java:2034]: Uhm - this method is typically used to set type-variables on wildcard types. It is a bit of a corner case in the compiler code, and I think it would be better to leave this alone. What is this code trying to do - and can the same be achieved in another way?

Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java:2034]: this code is preserving the reference projection of a type var, if it was originally declared as a reference `T.ref`, instead of overriding it with its dual `T`. If the reference projection is not preserved then it leads to information loss. I haven't found another more reliable way

Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java:2034]: OK I will do some more research here to see what alternatives we have, using this method is convenient as it is invoked whenever we do a mapping like for example from `Resolve::rawInstantiate` when we invoke `Types::memberType`, but I can try to find another way probably using a separate method to add the behavior I would like

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java [hunk 3]
```diff
@@ -1237,7 +1250,7 @@ public Boolean visitType(Type t, Type s) {
 
             private Set<TypePair> cache = new HashSet<>();
 
-            private boolean containsTypeRecursive(Type t, Type s) {
+            public boolean containsTypeRecursive(Type t, Type s) {
```
[the remark this answers is not in the document]
Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:1253]: sure makes sense

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java (hunk 1 above)
Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java]: yep you are right, fixed

**Comment by vicente-romero-oracle** [comment]
I have uploaded another commit, I think that it addresses all your comments so far, thanks a lot for the feedback!

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java [hunk 4]
```diff
@@ -1502,6 +1489,11 @@ public boolean isSignaturePolymorphic(MethodSymbol msym) {
      * Is t the same type as s?
      */
     public boolean isSameType(Type t, Type s) {
+        return isSameType(t, s, containsType);
+    }
+
+    public boolean isSameType(Type t, Type s, ContainsType containmentRelation) {
+        isSameTypeVisitor.setContainmentRel(containmentRelation);
```
Comment (mcimadamore) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:1496]: This should be reset after the call?

**Review by mcimadamore — COMMENTED** [review]
I think the pieces are all there. When looking at the code, the impression I have is that we're doing a lot of work just to pass some kind of boolean or enum flag which tells us if some extra loose behavior should be allowed or not (in both the subtyping, containment and sameType visitors).

I was about to suggest that in the previous review rounds, but then held back because the existing visitors cannot take external parameters. I think your changes to the `isSameType` visitor demonstrate that problem: the visitor is just a function on types, and there's no way to pass an extra boolean/enum to it - so you have to store it in a field, which then leads to issues because either (i) you create one visitor instance per subtyping check (which would be very expensive, I fear) or you resort to tricks using mutable fields.

Unfortunately fixing this mismatch is hard, as we're already using the visitor parameter to pass the second type of these type relations. So some other kind of workaround is required.

A possible suggestion for trying to get to a simpler form would be to introduce a new "parameterized" relation, like this:

```
public abstract static class ParameterizedTypeRelation<P> extends TypeRelation {
        P param;

        public final Boolean visit(Type t, Type s, P param) {
            P prevParam = this.param;
            try {
                this.param = param;
                return visit(t, s);
            } finally {
                this.param = prevParam;
            }
        }
    }
```

This basically generalizes the trick you did with `isSameType` (but resets the field value correctly after a visit). Then you make isSubtype/contains/isSameType subclasses of this visitor class, and define an enum with two values for the strict/loose behavior.

I believe this should give you enough flexibility to do what you want, and, at the same time, by minimizing set of changes compared to mainline.

File: src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java (hunk 4 above)
Comment (vicente-romero-oracle) [reply src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java:1496]: right

**Comment by vicente-romero-oracle** [comment]
> I think the pieces are all there. When looking at the code, the impression I have is that we're doing a lot of work just to pass some kind of boolean or enum flag which tells us if some extra loose behavior should be allowed or not (in both the subtyping, containment and sameType visitors).
> 
> I was about to suggest that in the previous review rounds, but then held back because the existing visitors cannot take external parameters. I think your changes to the `isSameType` visitor demonstrate that problem: the visitor is just a function on types, and there's no way to pass an extra boolean/enum to it - so you have to store it in a field, which then leads to issues because either (i) you create one visitor instance per subtyping check (which would be very expensive, I fear) or you resort to tricks using mutable fields.
> 
> Unfortunately fixing this mismatch is hard, as we're already using the visitor parameter to pass the second type of these type relations. So some other kind of workaround is required.
> 
> A possible suggestion for trying to get to a simpler form would be to introduce a new "parameterized" relation, like this:
> 
> ```
> public abstract static class ParameterizedTypeRelation<P> extends TypeRelation {
>         P param;
> 
>         public final Boolean visit(Type t, Type s, P param) {
>             P prevParam = this.param;
>             try {
>                 this.param = param;
>                 return visit(t, s);
>             } finally {
>                 this.param = prevParam;
>             }
>         }
>     }
> ```
> 
> This basically generalizes the trick you did with `isSameType` (but resets the field value correctly after a visit). Then you make isSubtype/contains/isSameType subclasses of this visitor class, and define an enum with two values for the strict/loose behavior.
> 
> I believe this should give you enough flexibility to do what you want, and, at the same time, by minimizing set of changes compared to mainline.

sure I will try this, thanks

**Comment by vicente-romero-oracle** [comment]
I have updated the PR with the refactoring you proposed, thanks

**Review by mcimadamore — APPROVED** [review]
Looks good - thanks for the hard work!

**Comment by vicente-romero-oracle** [comment]
> Looks good - thanks for the hard work!

nice, thanks to you for the feedback!
